flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala

```diff
@@ -355,4 +380,26 @@ object TableEnvironment {
     new ScalaStreamTableEnv(executionEnvironment, tableConfig)
   }
 
+  /**
+    * The primary workflow for executing plan validation for that generated from Table API.
+    * The validation is intentionally designed as a lazy procedure and triggered when we
+    * are going to run on Flink core.
+    */
+  class PlanPreparation(val env: TableEnvironment, val logical: LogicalNode) {
+
+    lazy val resolvedPlan: LogicalNode = env.getValidator.resolve(logical)
+
+    def validate(): Unit = env.getValidator.validate(resolvedPlan)
+
+    lazy val relNode: RelNode = {
+      env match {
+        case _: BatchTableEnvironment =>
```

Comment:
Yes, it shouldn't be distinguished. The difference should exist inside validator's rule set.